src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs

```diff
@@ -0,0 +1,332 @@
+                        packageReferenceArgs.PackageDependency.Id,
+                        packageReferenceArgs.ProjectPath));
+
+                    return 1;
+                }
+                else if (compatibleFrameworks.Count == restorePreviewResult.Result.CompatibilityCheckResults.Count())
+                {
+                    // Package is compatible with all the project TFMs
+                    // Add an unconditional package reference to the project
+                    packageReferenceArgs.Logger.LogInformation(string.Format(CultureInfo.CurrentCulture,
+                        Strings.Info_AddPkgCompatibleWithAllFrameworks,
+                        packageReferenceArgs.PackageDependency.Id,
+                        packageReferenceArgs.ProjectPath));
+
+                    // If the user did not specify a version then update the version to resolved version
+                    UpdatePackageVersionIfNeeded(restorePreviewResult, packageReferenceArgs);
+
+                    msBuild.AddPackageReference(packageReferenceArgs.ProjectPath,
+                        packageReferenceArgs.PackageDependency);
+                }
+                else
+                {
+                    // Package is compatible with some of the project TFMs
+                    // Add conditional package references to the project for the compatible TFMs
+                    packageReferenceArgs.Logger.LogInformation(string.Format(CultureInfo.CurrentCulture,
+                        Strings.Info_AddPkgCompatibleWithSubsetFrameworks,
+                        packageReferenceArgs.PackageDependency.Id,
+                        packageReferenceArgs.ProjectPath));
+
+                    var compatibleOriginalFrameworks = originalPackageSpec.RestoreMetadata
+                        .OriginalTargetFrameworks
+                        .Where(s => compatibleFrameworks.Contains(NuGetFramework.Parse(s)));
+
+                    // If the user did not specify a version then update the version to resolved version
+                    UpdatePackageVersionIfNeeded(restorePreviewResult, packageReferenceArgs);
+
+                    msBuild.AddPackageReferencePerTFM(packageReferenceArgs.ProjectPath,
+                        packageReferenceArgs.PackageDependency,
+                        compatibleOriginalFrameworks);
+                }
+            }
+            return 0;
+        }
+
+        private async Task<RestoreResultPair> PreviewAddPackageReference(PackageReferenceArgs packageReferenceArgs,
+            DependencyGraphSpec dgSpec,
+            PackageSpec originalPackageSpec)
+        {
+            // Set user agent and connection settings.
+            XPlatUtility.ConfigureProtocol();
+
+            var providerCache = new RestoreCommandProvidersCache();
+
+            using (var cacheContext = new SourceCacheContext())
+            {
+                cacheContext.NoCache = false;
+                cacheContext.IgnoreFailedSources = true;
```

Comment:
I would expect that this fails when a source is bad. This might accidentally ignore the source that has the highest version of the package and the user wouldn't know.